source/material_model/rheology/composite_visco_plastic.cc

```diff
@@ -82,38 +82,302 @@ namespace aspect
 
-            viscosity /= total_volume_fraction;
-            for (unsigned int j=0; j < 5; ++j)
-              partial_strain_rates[j] /= total_volume_fraction;
+        for (unsigned int i = 0; i < active_compositions.size(); ++i)
+          {
+            const double log_volume_fraction = std::log(volume_fractions[active_compositions[i]]);
+
+            if (use_diffusion_creep)
+              {
+                log_edot_and_deriv[i][0] = diffusion_creep->compute_log_strain_rate_and_derivative(log_viscoplastic_stress, pressure, temperature, grain_size, diffusion_creep_parameters[i]);
+                log_edot_and_deriv[i][0].first += log_volume_fraction;
+              }
+            if (use_dislocation_creep)
+              {
+                log_edot_and_deriv[i][1] = dislocation_creep->compute_log_strain_rate_and_derivative(log_viscoplastic_stress, pressure, temperature, dislocation_creep_parameters[i]);
+                log_edot_and_deriv[i][1].first += log_volume_fraction;
+              }
+            if (use_peierls_creep)
+              {
+                log_edot_and_deriv[i][2] = peierls_creep->compute_approximate_log_strain_rate_and_derivative(log_viscoplastic_stress, pressure, temperature, peierls_creep_parameters[i]);
+                log_edot_and_deriv[i][2].first += log_volume_fraction;
+              }
+            if (use_drucker_prager)
+              {
+                log_edot_and_deriv[i][3] = drucker_prager->compute_log_strain_rate_and_derivative(log_viscoplastic_stress, pressure, drucker_prager_parameters[i]);
+                log_edot_and_deriv[i][3].first += log_volume_fraction;
+              }
           }
-        return viscosity;
+
+        // 3) Calculate the total log strain rate from the first estimates
+        // for the component log strain rates.
+        // This will generally be more than the input total strain rate because
+        // the creep stress in Step 1 was been calculated assuming that
+        // only one mechanism was active, whereas the strain rate
+        // calculated in Step 2 allowed all the mechanisms to
+        // accommodate strain at that creep stress.
+        std::pair<double, double> log_edot_ii_and_deriv_iterate = calculate_isostress_log_strain_rate_and_derivative(log_edot_and_deriv,
+                                                                  viscoplastic_stress,
+                                                                  partial_strain_rates);
+        double log_strain_rate_residual = log_edot_ii_and_deriv_iterate.first - log_edot_ii;
+
+        // 4) In this rheology model, the total strain rate is partitioned between
+        // different flow components. We do not know how the strain is partitioned
+        // between these components.
+
+        // The following while loop contains a Newton iteration to obtain the
+        // viscoplastic stress that is consistent with the total strain rate.
```

Comment:
please add a comment saying that this also results in a correct partitioning between the flow mechanisms (the introductory sentences of the comment refer to the partitioning, but the last sentence doesnt explain that the Newton iteration will result in a correct partitioning).